### PR TITLE
Implement most of RFC 2930, providing the ReadBuf abstraction

### LINKED ISSUE #42788: Tracking issue for `Read::initializer`
Implemented in #42002.

### Open questions / blockers

* The code currently creates references to uninitialized integers (grep for `42788`), can we avoid that?

### PR DESCRIPTION
**Pull request by beepster4096**
This replaces the `Initializer` abstraction for permitting reading into uninitialized buffers, closing #42788.

This leaves several APIs described in the RFC out of scope for the initial implementation:

* read_buf_vectored
* `ReadBufs`

Closes #42788, by removing the relevant APIs.

### CODE REVIEW

**Comment by jyn514** [comment]
@DrMeepster do you mind squashing the commits? 41 is a *lot* to read through.

**Comment by DrMeepster** [comment]
Alright I squashed it.

**Comment by DrMeepster** [comment]
Hopefully, weird errors in files I didn't touch don't show up this time.

**Comment by DrMeepster** [comment]
uh I think this PR is cursed

**Comment by jyn514** [comment]
> uh I think this PR is cursed

We debugged this locally, here's an MCVE:

````
> cat doctest.rs
//! ```
//! extern crate proc_macro;
//!
//! # fn main() {}
//! ```
> rustdoc +rustc2 --test doctest.rs 

running 1 test
test doctest.rs - (line 1) ... FAILED

failures:

---- doctest.rs - (line 1) stdout ----
error: expected one of `;` or `as`, found `<eof>`
 --> doctest.rs:2:14
  |
2 | extern crate p
  |              ^ expected one of `;` or `as`
````

Notice how the error unexpectedly cuts off after `crate p`. So it's at least feasible it's related to this PR - either Rustdoc has UB or one of the changes introduced a logic error. It's strange that this only happens for rustdoc and not rustc.

**Comment by DrMeepster** [comment]
The `remove unresolved import` commit passing CI means that the error is caused by the changes to `BufReader`

**Comment by djc** [comment]
Very excited, thanks for working on this! I don't have a lot of experience with larger std changes like this, but I feel like it would be easier to land this if you focused on getting an initial PR merged that only includes the `ReadBuf` implementation (plus tests), and leave usage of the new type to future PRs (can still keep your changes, of course!).

**Comment by jyn514** [comment]
ping @sfackler, this is waiting on your review.

**Comment by DrMeepster** [comment]
When I fixed that error for windows, I really should have remembered to fix it for all the platforms :/

**Comment by jyn514** [comment]
ping @sfackler

**Comment by djc** [comment]
@joshtriplett as I understand it, @sfackler has removed himself from the reviewer list. Would you be able to take on reviewing this? Would be nice to make some progress on this important feature.

**Comment by JohnTitor** [comment]
Triage: @DrMeepster Could you resolve the merge conflict? Josh requested the perf run but the try build didn't happen because of it, I guess.

**Comment by DrMeepster** [comment]
Alright, the merge conflicts are resolved

**Comment by JohnTitor** [comment]
Another reason is that bors forgot about this PR as it's old :3

@bors try @rust-timer queue

**Comment by bstrie** [comment]
```
cargo-opt | avg: 0.9% | min: 0.0% | max: 3.6%
full | 502324728626 | 502561763352 | 0.0%
incr-full | 582142093056 | 582236028090 | 0.0%
incr-unchanged | 12931863153 | 12938818673 | 0.1%
incr-patched: println | 373247558921 | 386776372913 | 3.6%
```
I've never dug into perf output before, is this kind of behavior anomalous or typical, where one shows a big regression and all the rest show no difference?

**Comment by DrMeepster** [comment]
> ```
> cargo-opt | avg: 0.9% | min: 0.0% | max: 3.6%
> full | 502324728626 | 502561763352 | 0.0%
> incr-full | 582142093056 | 582236028090 | 0.0%
> incr-unchanged | 12931863153 | 12938818673 | 0.1%
> incr-patched: println | 373247558921 | 386776372913 | 3.6%
> ```
> I've never dug into perf output before, is this kind of behavior anomalous or typical, where one shows a big regression and all the rest show no difference?

That is really weird, not sure why only incr-patched would regress so hard.

**Comment by jyn514** [comment]
If the regressions are only in LLVM, then the codegen units the compiler picked are probably different. I wouldn't worry about it too much, there are a lot of false positives from CGUs.

**Comment by DrMeepster** [comment]
Can someone with perms rerun the perf then?

**Comment by jyn514** [comment]
@DrMeepster rerunning won't change the benchmarks. It's deterministic, it just varies wildly with small changes. https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/monomorphize/partitioning/index.html

**Comment by JohnCSimon** [comment]
Ping from triage: 
@DrMeepster Can you please address the merge conflicts. Thank you.

**Comment by DrMeepster** [comment]
Merge conflicts resolved

**Comment by DrMeepster** [comment]
I don't think this failure is related to my PR.

**Comment by DrMeepster** [comment]
> I don't think this failure is related to my PR.

well that's wrong, it seems I've accidentally changed Cargo.lock and a ton of submodules. No idea how I did that, probably messed up the last rebase somehow

**Comment by DrMeepster** [comment]
It looks like the merge conflicts were caused by my botched rebase too.

**Comment by djc** [comment]
@joshtriplett it's been three months since you've self-assigned this. What are the chances you'll have a chance to review it anytime soon? Would it be better to find another reviewer? It feels unfortunate that this is languishing when it can supposedly make a marked improvement to safely dealing with uninitialized memory (and help us on the road to a std `AsyncRead` trait).

**Comment by joshtriplett** [comment]
Sorry for the delay. (It's been about one month since this was ready for review. Sorry that I haven't gotten to it in that time.)

**Comment by joshtriplett** [comment]
@bors r+

This looks great. It needs another round of review when it comes up for stabilization, but in the meantime, this looks reasonable for experimentation in nightly.

File: library/std/src/io/mod.rs [hunk 1]
```diff
@@ -810,7 +796,40 @@ pub trait Read {
...
+        while buf.remaining() > 0 {
+            let prev_filled = buf.filled().len();
+            match self.read_buf(buf) {
+                Ok(()) => {}
+                Err(e) if e.kind() == ErrorKind::Interrupted => continue,
+                Err(e) => return Err(e),
+            }
+
+            if buf.filled().len() == prev_filled {
+                return Err(Error::new(ErrorKind::UnexpectedEof, "failed to fill buffer"));
```
Comment (klensy) [inline library/std/src/io/mod.rs:825]: This can be `new_const`?

File: library/std/src/sys/unsupported/fs.rs [hunk 2]
```diff
@@ -206,6 +206,10 @@ impl File {
         self.0
     }
 
+    pub fn read_buf(&self, buf: &mut ReadBuf<'_>) -> io::Result<()> {
```
Comment (joshtriplett) [inline library/std/src/sys/unsupported/fs.rs:209]:
```suggestion
    pub fn read_buf(&self, _buf: &mut ReadBuf<'_>) -> io::Result<()> {
```

**Comment by jyn514** [comment]
```
  Downloading crates ...
  Downloaded byteorder v1.4.3
thread 'main' panicked at 'range end index 8192 out of range for slice of length 1804', /checkout/library/std/src/io/readbuf.rs:96:58
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'tests failed for https://github.com/diesel-rs/diesel', src/tools/cargotest/main.rs:101:9
```
Something is very wrong, this failed for rustdoc too: https://github.com/rust-lang/rust/pull/81156#issuecomment-763173021

**Comment by DrMeepster** [comment]
Interesting, I fixed it for rustdoc that time. This must be an unrelated similar error. I'll look into it

**Comment by DrMeepster** [comment]
I have been unable to find the source of the failure still :/

**Comment by gilescope** [comment]
Linus's law is the assertion that "given enough eyeballs, all bugs are shallow". Let's test this.

**Comment by gilescope** [comment]
Could `ReadBufs` be implemented as just doing the default of taking the first buffer and treating that buffer in the exact same way as read_buf? Other specialisations could provide more efficient implementations than the default.

**Comment by gilescope** [comment]
How goes? Is this potentially ready to be merged in to nightly?

File: library/std/src/io/readbuf.rs [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    ///
+    /// # Panics
+    ///
+    /// Panics if `self.remaining()` is less than `n`.
+    #[inline]
+    pub fn initialize_unfilled_to(&mut self, n: usize) -> &mut [u8] {
+        assert!(self.remaining() >= n);
+
+        let extra_init = self.initialized - self.filled;
+        // If we dont have enough initialized, do zeroing
```
Comment (joshtriplett) [inline library/std/src/io/readbuf.rs:139]:
```suggestion
        // If we don't have enough initialized, do zeroing
```

(and in various other places; not a blocker for merging, just a nit)

File: library/std/src/io/readbuf.rs [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        //SAFETY: We only slice the filled part of the buffer, which is always valid
+        unsafe { MaybeUninit::slice_assume_init_ref(&self.buf[0..self.filled]) }
+    }
+
+    /// Returns a mutable reference to the filled portion of the buffer.
+    #[inline]
+    pub fn filled_mut(&mut self) -> &mut [u8] {
+        //SAFETY: We only slice the filled part of the buffer, which is always valid
+        unsafe { MaybeUninit::slice_assume_init_mut(&mut self.buf[0..self.filled]) }
+    }
```
[the remark this answers is not in the document]
Comment (joshtriplett) [reply library/std/src/io/readbuf.rs:79]: It appears to be covered now.

**Comment by joshtriplett** [comment]
I just read through this again in detail. There's one outstanding comment regarding a method name in the public API (`assume_init` vs something like `add_init` or similar that makes it clear that it's adding to the initialized region rather than replacing it).

That issue aside, however, I think this implementation is ready for a T-libs-api FCP to confirm. (As far as I can tell, parts of this are insta-stable, and thus need an FCP.)

@rfcbot merge

**Comment by joshtriplett** [comment]
@rfcbot concern Confirm API naming for assume_init

**Comment by joshtriplett** [comment]
In the interests of collecting possible naming considerations in one place:

- The RFC uses `assume_init`.
- `assume` seems evocative of an `unsafe` API: you're not *making* something initialized, you're making an assumption.
- `add` or similar would more clearly indicate the addition to the initialized region, rather than replacing it.
- Something like `assume_added_init` or `add_assumed_init` would have the benefits of both, but be more verbose.

**Comment by nrc** [comment]
> As far as I can tell, parts of this are insta-stable, and thus need an FCP

@joshtriplett which parts will be insta-stable?

**Comment by joshtriplett** [comment]
We reviewed this in today's @rust-lang/libs-api meeting, and confirmed that there are no insta-stable bits of this. We *do* see two potential issues with this that should be covered and addressed in the tracking issue:

- [`assume_init` naming](https://github.com/rust-lang/rust/pull/81156#issuecomment-988536835)
- `&mut ReadBuf` allows replacing the whole ReadBuf, which would allow pointing it to a different buffer of memory. Some callers may not expect that, and that may be error-prone for some high-performance use cases. We should consider having an opaque type wrapping a `&mut ReadBuf` (name subject to bikeshedding), adding a method to `ReadBuf` to return one, and then using that as the argument to `read_buf`.

However, since this isn't insta-stable, we can address both of those through further development on nightly.

@rfcbot cancel

@bors r+

**Comment by Mark-Simulacrum** [comment]
FWIW, I think it would be nice to squash the commits here, but it doesn't seem worth blocking over and it looks like the feature letting upstream maintainers push to the branch isn't turned on for this PR.

File: library/std/src/io/copy.rs [hunk 5]
```diff
@@ -82,33 +82,30 @@ impl<I: Write> BufferedCopySpec for BufWriter<I> {
...
         }
 
-        // FIXME: #42788
-        //
-        //   - This creates a (mut) reference to a slice of
-        //     _uninitialized_ integers, which is **undefined behavior**
-        //
-        //   - Only the standard library gets to soundly "ignore" this,
-        //     based on its privileged knowledge of unstable rustc
-        //     internals;
```
Comment (RalfJung) [inline library/std/src/io/copy.rs:92]: Ah, I am so happy to see this comment disappear. :)

**Pull request merged**